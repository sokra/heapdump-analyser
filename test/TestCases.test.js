const { exec, fork } = require("child_process");
const fs = require("fs");
const path = require("path");
const rootPath = path.resolve(__dirname, "..");
const findLeaks = path.resolve(rootPath, "find-leaks.js");

const cleanOutput = (output) => {
	return output
		.replace(/@\d+/g, "@XXX")
		.replace(/internal \d+/g, "internal XXX");
};

describe("TestCase", () => {
	const testCasesPath = path.resolve(__dirname, "cases");
	for (const caseName of fs.readdirSync(testCasesPath)) {
		if (caseName.startsWith(".")) continue;
		describe(caseName, () => {
			it("should run the case", (done) => {
				exec(
					"node index.js",
					{
						cwd: path.resolve(testCasesPath, caseName),
					},
					done
				);
			});
			it("should analyse the result", (done) => {
				const file = fs.readFileSync(
					path.resolve(testCasesPath, caseName, "index.js")
				);
				const match = /^\/\/\s+(.+)\n/.exec(file);
				expect(match).toBeTruthy();
				const process = fork(
					findLeaks,
					[
						path.resolve(testCasesPath, caseName, "test.heapsnapshot"),
						match[1],
					],
					{
						cwd: rootPath,
						stdio: ["inherit", "pipe", "inherit", "ipc"],
					}
				);
				let output = "";
				process.stdout.on("data", (chunk) => (output += chunk));
				process.on("exit", (code) => {
					expect(code).toBe(0);
					expect(cleanOutput(output)).toMatchSnapshot();
					done();
				});
			});
		});
	}
});
