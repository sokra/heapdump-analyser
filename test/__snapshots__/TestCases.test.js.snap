// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestCase cycle should analyse the result 1`] = `
"     synthetic (Stack roots) @XXX
    internal X = object system / Context @XXX
   context GLOBAL = object { maps, key } @XXX
+ .key = object TestObject #1 @XXX


    object TestObject #1 @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #2 @XXX
 |       synthetic (Stack roots) @XXX
 |      internal X = object system / Context @XXX
 |     context GLOBAL = object { maps, key } @XXX
 |    .maps = object Array @XXX
 |   [0] = object WeakMap @XXX
 |  internal table = array  @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #2 @XXX
  object TestObject #2 @XXX


         synthetic (Stack roots) @XXX
        internal X = object system / Context @XXX
       context GLOBAL = object { maps, key } @XXX
      .maps = object Array @XXX
     [1] = object WeakMap @XXX
    internal table = array  @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #3 @XXX
 |  object TestObject #2 @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #3 @XXX
  object TestObject #3 @XXX


    object TestObject #3 @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #4 @XXX
 |          synthetic (Stack roots) @XXX
 |         internal X = object system / Context @XXX
 |        context GLOBAL = object { maps, key } @XXX
 |       .maps = object Array @XXX
 |      [0] = object WeakMap @XXX
 |     internal table = array  @XXX
 |  + internal X / part of key (TestObject @XXX) -> value (WeakMap @XXX) pair in WeakMap (table @XXX) = object WeakMap @XXX
 |  |  object TestObject #2 @XXX
 |  + internal X / part of key (TestObject @XXX) -> value (WeakMap @XXX) pair in WeakMap (table @XXX) = object WeakMap @XXX
 |   object WeakMap @XXX
 |  internal table = array  @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #4 @XXX
  object TestObject #4 @XXX
"
`;

exports[`TestCase prefer-easier-path should analyse the result 1`] = `
"      synthetic (Stack roots) @XXX
     internal X = object system / Context @XXX
    context GLOBAL = object { map, key, deep } @XXX
   .deep = object { nested } @XXX
+ .nested = object TestObject #1 @XXX
"
`;

exports[`TestCase unreferenced should analyse the result 1`] = `""`;

exports[`TestCase weak-map should analyse the result 1`] = `
"     synthetic (Stack roots) @XXX
    internal X = object system / Context @XXX
   context GLOBAL = object { map, key, deep } @XXX
+ .key = object TestObject #1 @XXX


       synthetic (Stack roots) @XXX
      internal X = object system / Context @XXX
     context GLOBAL = object { map, key, deep } @XXX
    .deep = object { nested } @XXX
   .nested = object { key, map } @XXX
+ .key = object TestObject #2 @XXX


        synthetic (Stack roots) @XXX
       internal X = object system / Context @XXX
      context GLOBAL = object { map, key, deep } @XXX
     .map = object WeakMap @XXX
    internal table = array  @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #3 @XXX
 |  object TestObject #2 @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #3 @XXX
  object TestObject #3 @XXX


    object TestObject #1 @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #4 @XXX
 |        synthetic (Stack roots) @XXX
 |       internal X = object system / Context @XXX
 |      context GLOBAL = object { map, key, deep } @XXX
 |     .deep = object { nested } @XXX
 |    .nested = object { key, map } @XXX
 |   .map = object WeakMap @XXX
 |  internal table = array  @XXX
 + internal X / part of key (TestObject @XXX) -> value (TestObject @XXX) pair in WeakMap (table @XXX) = object TestObject #4 @XXX
  object TestObject #4 @XXX
"
`;
